lib/watir/locators/button/selector_builder/xpath.rb

```diff
@@ -3,68 +3,57 @@ module Locators
     class Button
       class SelectorBuilder
         class XPath < Element::SelectorBuilder::XPath
-          def build(selector)
-            return super if selector.key?(:adjacent)
-
-            selector[:tag_name] = 'button'
-
-            type = selector.delete :type
-            return super(selector) if type.eql?(false)
-
-            # both :value and :text selectors will locate elements by value attribute or text content
-            selector[:text] = selector.delete(:value) if selector.key?(:value)
-
-            wd_locator = super(selector)
-
-            start_string = default_start
-
-            button_string = "local-name()='button'"
-            common_string = wd_locator[:xpath].gsub(start_string, '').gsub("[#{button_string}]", '')
-
-            input_string = "(local-name()='input' and #{input_types(type)})"
-
-            tag_string = if type.nil?
-                           "[#{button_string} or #{input_string}]"
-                         else
-                           "[#{input_string}]"
-                         end
+          private
 
-            xpath = "#{start_string}#{tag_string}#{common_string}"
+          def tag_string
+            return super if @adjacent
 
-            {xpath: xpath}
-          end
+            @selector.delete(:tag_name)
 
-          protected
+            type = @selector.delete :type
 
-          def use_index?
-            false
+            if type
```

Comment:
The type attribute is valid for `button` and `input` elements, so we cannot assume a tag name based on its presence.

For example, given:

~~~~~~~~
<button id="btn_type" type="submit">text</button>
<input id="id_type" type="submit" value="text">
~~~~~~~~

With the new changes:

~~~~~~~~
browser.buttons(type: 'submit').map(&:id)
#=> ["id_type"]
~~~~~~~~

In 6.14:

~~~~~~~~
browser.buttons(type: 'submit').map(&:id)
#=> ["btn_type", "id_type"]
~~~~~~~~

Reply:
grrr, good catch. Thanks.